File: DATN-Fashion-Shop-BE/src/main/java/com/example/DATN_Fashion_Shop_BE/model/Promotion.java
package com.example.DATN_Fashion_Shop_BE.model;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.envers.Audited;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "promotions")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Audited
public class Promotion extends BaseEntity{
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "discount_percentage", precision = 5, nullable = false)
    private Double discountPercentage;

    @Column(name = "descriptions", columnDefinition = "NVARCHAR(MAX)")
    private String descriptions;

    @Column(name = "start_date", nullable = false)
    private LocalDateTime startDate;

    @Column(name = "end_date", nullable = false)
    private LocalDateTime endDate;

    @Column(name = "is_active", nullable = false)
    private Boolean isActive;

    @OneToMany(mappedBy = "promotion", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.EAGER)
    private Set<Product> products = new HashSet<>();
}

File: FE/DATN-Fashion-Shop-FE/src/app/models/Product/Promotion.ts
import { BaseEntity} from '../BaseEntity';
import { Product } from './product';

export class Promotion extends BaseEntity {
  id: number;
  discountPercentage: number;
  description?: string;
  startDate: Date;
  endDate: Date;
  isActive: boolean;
  products: Product[];

  constructor(data?: Partial<Promotion>) {
    super(data);
    this.id = data?.id || 0;
    this.discountPercentage = data?.discountPercentage || 0;
    this.description = data?.description || '';
    this.startDate = data?.startDate ? new Date(data.startDate) : new Date();
    this.endDate = data?.endDate ? new Date(data.endDate) : new Date();
    this.isActive = data?.isActive ?? false;
    this.products = data?.products?.map(p => new Product(p)) || [];
  }

  isPromotionActive(): boolean {
    const now = new Date();
    return this.isActive && now >= this.startDate && now <= this.endDate;
  }
}
